JsonArray and transformations API implementation

kasiabulat opened the pull request:
I implemented `JsonArray` and transformations API for `JsonNode`. In order to do this, I needed to adjust other types as well. I also made some minor fixes and included some feedback from discussions. I still new to add much more tests and documentation to some of the methods, but I wanted to create this PR today to discuss new issues that came up during tomorrow meeting. I can split it in to couple of PRs later on, if the number of changes it too big there.

New open questions (I added them to the open questions section in specification as well):
* Where do we want to enable user to set handling properties manner? Do we want to support it as `JsonElement` does not? (and currently `JsonNode.Parse` implementation uses `JsonDocument.Parse`)
* Do we want to support transforming `JsonObject` into JSON string? Should `ToString` behave this way or do we want an additional method - e.g. `GetJsonString` (it might be confusing as we have a type `JsonString`) or `GetJsonRepresenation`?
* `AsJsonElement` function on `JsonNode` currently does not work for `null` node. Do we want to support null case somehow?
* Do we want separate `JsonElement` implementations for `JsonNode` and `JsonDocument` or can the implementation be mixed as it is now (with a lot of ifs in each method)?
* Do we want separate `JsonElement.ArrayEnumerator` and `JsonElement.ObjectEnumerator` implementations for `JsonNode` and `JsonDocument`? Do we want separate methods in `JsonElement` returning strongly typed different implementations?
* Is it OK that `JsonElement.GetRawText` does not return the raw text for `JsonNode` parent?
* Should we support `JsonElement.WriteTo` for `JsonNode` parent?
* Private property `JsonElement.TokenType` is currently throwing an `InvalidCast` exception for `JsonNode` parent in debugger. Is it OK?
* Do we want both `Clone` and `DeepCopy` methods for `JsonNode`? 
* Are we OK with needing to cast `JsonArray` to `JsonNode` in order to add it to `JsonObject`? (there's an ambiguous call between `JsonArray` and `IEnumerable<JsonNode>` right now)
* Do we want `IsImmutable` property for `JsonElement`?
* Is `DateTime` for `JsonString` handled correctly?

addresses: #39922
cc: @joperezr @bartonjs @ericstj @ahsonkhan @terrajobst @JamesNK @stephentoub

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,668 @@
+        {
+            foreach (string value in values)
+            {
+                _list.Add(new JsonString(value));
```
NIT: I believe @stephentoub already gave you this feedback before, but in here one of the values might be null, which theoretically that's fine but it is weird that if you right today `var str = new JsonString();` that will set the value to string.empty.

kasiabulat replied:
I think `JsonString` shouldn't accept null, as for such cases we have null `JsonNode`. In here, JsonString constructor will throw an exception. 

Previous comment you mentioned: https://github.com/dotnet/corefx/pull/40107#discussion_r311722058

joperezr replied:
right, but if that's the case, I probably would expect that `var string = new JsonString();` should return me a new JsonString object which has null as value

bartonjs replied:
A `null` string in the `values` should produce a JSON null in the output.

kasiabulat replied:
Ok, I will fix it.

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,668 @@
+        /// </exception>
+        public JsonNode this[int idx]
+        {
+            get => 0 <= idx && idx < _list.Count ? _list[idx] : throw new ArgumentOutOfRangeException();
```
>0 [](start = 19, length = 1)

NIT: can we add parenthesis around the condition? having a `=> 0 <=` is not super readable.

bartonjs replied:
Or just make it a statement-bodied member and write the braces, if, and return.  (Much more legible, IMO)

kasiabulat replied:
Right, sure, I'll change it.

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,668 @@
+        /// </exception>
+        public JsonNode this[int idx]
+        {
+            get => 0 <= idx && idx < _list.Count ? _list[idx] : throw new ArgumentOutOfRangeException();
```
NIT: Also, should we add a message to the exception?

JamesNK commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -13,11 +13,113 @@ public enum DuplicatePropertyNameHandling
+        public bool IsReadOnly { get { throw null; } }
+        public System.Text.Json.JsonNode this[int idx] { get { throw null; } set { } }
+        public override System.Text.Json.JsonValueKind ValueKind { get { throw null; } }
+        public void Add(bool value) { }
```
Are all these overloads necessary? I think `JsonNode` has implicit conversions from primitive types. If that is the case calling `Add(JsonNode value)` with `Add("hello world")` will do the work for you.

JamesNK replied:
Same question for ctor with IEnumerables, and Insert overloads.

kasiabulat replied:
All primitive types have implicit conversions, but `JsonNode` itself does not, so it wouldn't work without overloads.

joperezr replied:
Also, I believe the point of having all the overloads is so that you can use collection instantiation like:

```c#
var jsonObject = new JsonObject
{
  "SomeBool": true,
  "SomeString": "hello",
  "SomeNumber": 56.7,
  "SomeArray": new JsonArray
  {
    { false, "goodbye", 23 }
  }
};
```
We really only need one new statement above given that we  have all the overloads for JsonObject and JsonString.

kasiabulat replied:
I think James meant that it all (including collection instantiation) should work even without overloads. But implicit conversions on JsonString, JsonNumber and JsonBoolean aren't sufficient. As I remember, at the beginning we tried the approach with implicit operators on JsonNode as well, but it seemed to require too many hacks to make it do all the magic (and it might have been too magical for users).

JamesNK replied:
What hacks are required? Implicit conversions on JToken has worked well for Newtonsoft.Json

kasiabulat replied:
I found it - the hacks were needed because we had a JsonNode(Array) implicit conversion, and we needed to check if we want to create JsonObject or JsonArray from it. But in current version we anyway resigned from Array. The main downside of implicit operators would be then the magical behavior for users - they may experience conversion happening in places they wouldn't expect. And most of them would not know what is happening. Quoting @joperezr : "there are more places that expose JsonNode than more concrete types, so basically having implicit operators on JsonNode for everything might prove that we end up boxing things we don't really intend to".

JamesNK commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -178,6 +281,19 @@ public abstract partial class JsonNamingPolicy
+        public static System.Text.Json.JsonNode DeepCopy(System.Text.Json.JsonDocument jsonDocument) { throw null; }
+        public static System.Text.Json.JsonNode DeepCopy(System.Text.Json.JsonElement jsonElement) { throw null; }
+        public static System.Text.Json.JsonNode DeepCopy(System.Text.Json.JsonNode jsonNode) { throw null; }
+        public static System.Text.Json.JsonNode GetNode(System.Text.Json.JsonElement jsonElement) { throw null; }
```
Do changes in the node get mirrored in the element?

If not then a name like CreateNode would be better to indicate that there is a copy here. Likewise, AsJsonElement should be CreateJsonElement.

kasiabulat replied:
Thank you for pointing this out! In current version changes are mirrored, so naming should be appropriate. But we can discuss what behavior do we want.

JamesNK commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArrayEnumerator.cs:
```diff
@@ -0,0 +1,45 @@
+    /// <summary>
+    ///   Supports an iteration over a JSON array.
+    /// </summary>
+    public struct JsonArrayEnumerator : IEnumerator<JsonNode>
```
readonly?

bartonjs replied:
It isn't readonly, it modifies via MoveNext...

JamesNK replied:
The internal `IEnumerator<JsonNode>` is modified, but the reference to that is always readonly, no?

bartonjs replied:
That's true for the current implementation; but if the enumerator were lifted more (including correcting the field type to be the List<T>.Enumerator struct-enumerator) then mutable-struct semantics would apply.

Since the purpose of the enumerator is to modify I think the readonly would be a bad signal.  Pass-through properties for reference types clearly make things a little murky, but I would argue against putting `readonly` as a modifier on the MoveNext method since it makes the struct now represent a different state; and (IIRC) the readonly modifier on the type just means "apply it to all the members".

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,668 @@
+        ///   Returns an enumerator that iterates through the JSON array values.
+        /// </summary>
+        /// <returns>An enumerator structure for the <see cref="JsonArray"/>.</returns>
+        IEnumerator IEnumerable.GetEnumerator() => _list.GetEnumerator();
```
any reason why we wanted to implement this method explicitly instead of implicitly?>

kasiabulat replied:
It is required by `IList` interface. And this method should of course return JsonArrayEnumerator, I will fix it.

bartonjs replied:
GetEnumerator() is defined publicly (next method) as returning `IEnumerator<JsonNode>`.  Since `IEnumerable<T>` extends `IEnumerable` with a conflicting definition of `GetEnumerator()` at least one of the two of them always has to be explicitly implemented.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,675 @@
+        }
+
+        /// <summary>
+        ///   Adds the specified <see cref="JsonNode"/> value to the JSON array.
```
99% of the instances of "JSON" in the comments here don't really apply.  Instead of "the JSON array" here I'd say "the collection".  It may have been a JSON array value during Parse, and it may be one again during WriteTo, but for the purposes of Add it's just a collection.  You could also say "as the last item in this collection", since you're an IList (and thus ordered).

Our current docs style guides say to not really reference the current type name in method summaries.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,675 @@
+        ///   Adds the specified <see cref="JsonNode"/> value to the JSON array.
+        /// </summary>
+        /// <param name="value">The value to add.</param>
+        /// <remarks>Null value is allowed and represents a null JSON node.</remarks>
```
```suggestion
        /// <remarks>Null value is allowed and represents the JSON null value.</remarks>
```

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,675 @@
+        /// </summary>
+        /// <param name="value">The value to add.</param>
+        /// <exception cref="ArgumentNullException">
+        ///   Provided value is null.
```
Nulls should work here, producing the JSON null value.

kasiabulat replied:
Ok, I will fix it.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,675 @@
+        /// </summary>
+        /// <param name="index">The zero-based index at which item should be inserted.</param>
+        /// <param name="item">The item to add.</param>
+        /// <remarks>Null item is allowed and represents a null JSON node.</remarks>
```
```suggestion
        /// <remarks>The <paramref name="item"/> parameter may be <see langword="null" />, which represents the JSON null value.</remarks>
```

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,675 @@
+        /// <summary>
+        ///   Inserts the specified item at the specified index of the JSON array.
+        /// </summary>
+        /// <param name="index">The zero-based index at which item should be inserted.</param>
```
```suggestion
        /// <param name="index">The zero-based index at which <paramref name="item"/> should be inserted.</param>
```

(applies to all the overloads)

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,675 @@
+        ///   <see langword="true"/> if the value is successfully found in a JSON array,
+        ///   <see langword="false"/> otherwise.
+        /// </returns>
+        public bool Contains(JsonNode value) => _list.Contains(value);
```
All of these Contains methods are going to be hard to describe.  I recommend removing them.

(Is it reference equality, or value equality? What does it mean to Contains(5.0f)? `5`? `5.0`? `5.0000000000000000000000000000000000000000000000000000000001`? `51e-1`?

kasiabulat replied:
:disappointed_relieved:

bartonjs replied:
Given that this method (and IndexOf) are doing reference equality, we should delete Contains. IndexOf, and LastIndexOf until we can come up with a stronger definition.

This is made worse by the implicit conversions, because  `array.Add(true)`, `array.Contains(true) => false`.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,675 @@
+        ///   Returns an enumerator that iterates through the JSON array values.
+        /// </summary>
+        /// <returns>An enumerator structure for the <see cref="JsonArray"/>.</returns>
+        public IEnumerator<JsonNode> GetEnumerator() => new JsonArrayEnumerator(this);
```
```suggestion
        public JsonArrayEnumerator GetEnumerator() => new JsonArrayEnumerator(this);
```

That might also require an explicit interface implementation for one returning `IEnumerator<JsonNode>`.  I don't remember.

kasiabulat replied:
Ok, I will check it.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArray.cs:
```diff
@@ -0,0 +1,675 @@
+        ///   Creates a new JSON array that is a copy of the current instance.
+        /// </summary>
+        /// <returns>A new JSON array that is a copy of this instance.</returns>
+        public override JsonNode Clone() => new JsonArray(_list);
```
So a shallow clone?  Is that what we want?

kasiabulat replied:
This should call `JsonArray(IEnumerable<JsonNode> values)`, which should call `new List<JsonNode>(values)`, which according to documentation should "initialize a new instance of the List<T> class that contains elements copied from the specified collection".

bartonjs replied:
```C#
JsonArray inner = new JsonArray { 1, 2, 3 };
JsonArray outer = new JsonArray { (JsonNode)inner) };
JsonArray outerClone = (JsonArray)outer.Clone();
((JsonArray)outerClone[0]).Add(4);

int innerLength = inner.Length;
```

This code will produce 4.  A deep clone would produce 3.  The difference is that while the list in `outer` got copied (so adding to outer doesn't affect outerClone) it didn't copy the things inside the list, so the mutable `inner` value is now in both of them.

kasiabulat replied:
I see, thanks!

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.cs:
```diff
@@ -2,6 +2,12 @@
 // See the LICENSE file in the project root for more information.
 
+// for now disabling error caused by not adding documentation to methods
+#pragma warning disable CS1591
```
It's easier to add them now. (Especially if only one Parse is supported in this PR).

kasiabulat replied:
You mean add documentation? Yes, sure, I am planning to, I just didn't have time yet.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.cs:
```diff
@@ -10,5 +16,105 @@ namespace System.Text.Json
+                case JsonValueKind.Null:
+                    return null;
+                default:
+                    throw new ArgumentException();
```
Debug.Assert(jsonElement.ValueKind == JsonValueKind.Unknown, "No handler for JsonValueKind.{jsonElement.ValueKind}");

bartonjs replied:
ArgumentExceptions should have a message and identify the parameter by name via the `paramName` constructor argument.  (There's probably one somewhere about `default(JsonElement)`, which is the only case that can hit this.)

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -475,6 +496,51 @@ public bool TryGetJsonObjectPropertyValue(string propertyName, out JsonObject js
+                return jsonArray;
+            }
+
+            throw new InvalidCastException(SR.Format(SR.PropertyTypeMismatch, propertyName));
```
Don't throw InvalidCastException manually.  This should presumably be an ArgumentException.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs:
```diff
@@ -475,6 +496,51 @@ public bool TryGetJsonObjectPropertyValue(string propertyName, out JsonObject js
+            {
+                if (jsonNode is JsonArray jsonNodeCasted)
+                {
+                    jsonArray = jsonNodeCasted;
```
```C#
        public bool TryGetJsonArrayPropertyValue(string propertyName, out JsonArray jsonArray)
        {
            if (TryGetPropertyValue(propertyName, out JsonNode jsonNode))
            {
                jsonArray = jsonNode as JsonArray;
                return jsonArray != null;
            }

            jsonArray = null;
            return false;
        }
```

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonString.cs:
```diff
@@ -38,16 +40,28 @@ public sealed class JsonString : JsonNode, IEquatable<JsonString>
         /// <param name="value">The value to represent as a JSON string.</param>
-        public JsonString(DateTimeOffset value) => Value = value.ToString();
+        /// <exception cref="FormatException">
+        ///   Text value of this instance is not in an ISO 8601 defined DateTimeOffset format.
```
Why is the ctor throwing a FormatException?

kasiabulat replied:
Right, it shouldn't be.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonString.cs:
```diff
@@ -67,6 +81,70 @@ public string Value
+        /// <exception cref="FormatException">
+        ///   Text value of this instance is not in a GUID recognized format.
+        /// </exception>
+        public Guid GetGuid() => Guid.Parse(_value);
```
This supports all GUID formats, the rest of the JSON stack only supports the "D" format.  This should be restricted, too.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonString.cs:
```diff
@@ -67,6 +81,70 @@ public string Value
+        /// <exception cref="FormatException">
+        ///   Text value of this instance is not in a GUID recognized format.
+        /// </exception>
+        public bool TryGetGuid(out Guid value) => Guid.TryParse(_value, out value);
```
Restrict this to the "D" format. (TryParseExact)

bartonjs replied:
GetGuid got changed to D, but TryGetGuid did not.

kasiabulat commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -178,6 +281,19 @@ public abstract partial class JsonNamingPolicy
+        public abstract System.Text.Json.JsonNode Clone();
+        public static System.Text.Json.JsonNode DeepCopy(System.Text.Json.JsonDocument jsonDocument) { throw null; }
+        public static System.Text.Json.JsonNode DeepCopy(System.Text.Json.JsonElement jsonElement) { throw null; }
+        public static System.Text.Json.JsonNode DeepCopy(System.Text.Json.JsonNode jsonNode) { throw null; }
```
Currently `JsonNode.DeepCopy(node)` and `node.Clone()` are the same and I wondered if we need both. Yes, sure, I should make `DeepCopy(JsonElement)` check if the parent is JsonNode.

JamesNK replied:
Get rid of clone IMO. No one knows what clone means (cloning the reference, cloning the current type, cloning the entire hierarchy). DeepCopy is fine by itself.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Document/JsonElement.ArrayEnumerator.cs:
```diff
@@ -19,19 +19,49 @@ public struct ArrayEnumerator : IEnumerable<JsonElement>, IEnumerator<JsonElemen
+                    _endIdx = -1;
+
+                    var jsonArray = (JsonArray)target._parent;
+                    _jsonArrayEnumerator = new JsonArrayEnumerator(jsonArray);
```
Ok, thanks. I am not sure where would I set "modification iteration"?

bartonjs replied:
it's a thing List does.  Every time you call Add/Remove/Clear/Insert (and probably the positional assignment) it does `_version++`; then the enumerators throw in MoveNext if their copy of _version disagrees.  Unless _curIdx is being used as a sentinel to understand which mode we're in, you'd just save (your own copy of) jsonArray._version (which you'd have to define) to _curIdx, and compare it in MoveNext.

Or we can make the struct bigger by wrapping the other struct... but that doesn't feel necessary.

kasiabulat replied:
Ok, I can use my own "_version", thank you for explanation!

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs:
```diff
@@ -27,15 +27,50 @@ internal JsonElement(JsonDocument parent, int idx)
+        {
+            get
+            {
+                JsonDocument document = (JsonDocument)_parent;
```
Ok, but should I support this property for `JsonNode` parent? (it shouldn't be called in that case)

bartonjs replied:
I'm solving the "I see exceptions in the debugger" problem :)

kasiabulat replied:
Ok, thanks, I can change it to `as`. But you would suggest to implement this method for JsonNode as well, though it shouldn't be used (and maybe add debug assert)? Or keep it as it is just with changing casting?

bartonjs replied:
A Debug.Assert won't help either (it's worse, actually, the debugger display probably won't trap it).  You'd have to put an attribute on the property to not display in the debugger, which is fine to do. 

Actually, that's better than the `as` since it'll identify if you call it from the wrong scope.

```C#
[DebuggerBrowsable(DebuggerBrowsableState.Never)]
```

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs:
```diff
@@ -228,7 +283,9 @@ public JsonElement GetProperty(ReadOnlySpan<byte> utf8PropertyName)
             if (propertyName == null)
+            {
                 throw new ArgumentNullException(nameof(propertyName));
+            }
```
Ok, I thought it might have been overlooked ;)

JamesNK commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArrayEnumerator.cs:
```diff
@@ -0,0 +1,45 @@
+        ///   Initializes a new instance of the <see cref="JsonArrayEnumerator"/> class supporting an interation over provided JSON array.
+        /// </summary>
+        /// <param name="jsonArray">JSON array to iterate over.</param>
+        public JsonArrayEnumerator(JsonArray jsonArray) => _enumerator = jsonArray._list.GetEnumerator();
```
This boxes the struct enumerator from list into `IEnumerator<JsonNode>`?

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonArrayEnumerator.cs:
```diff
@@ -0,0 +1,45 @@
+    /// </summary>
+    public struct JsonArrayEnumerator : IEnumerator<JsonNode>
+    {
+        private readonly IEnumerator<JsonNode> _enumerator;
```
Should/can this be `private readonly List<JsonNode>.Enumerator _enumerator;`?

kasiabulat replied:
I can change it. Why would this be better?

bartonjs replied:
List returns a struct-based enumerator.  This is a struct-based enumerator.  If you store `IEnumerator<JsonNode>` then you have to pay the GC cost of boxing and the runtime cost of interface virtual dispatch.  If the answer is that it's always the List struct-based enumerator then we can keep the optimization (and perhaps have to change it later).

Really it's that this whole type only makes sense if you're avoiding boxing the list enumerator; otherwise you'd just return it directly.

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+
+            jsonArray.Add("value");
+            Assert.Equal("value", (JsonString)jsonArray[idx++]);
+            Assert.True(jsonArray.Contains("value"));
```
```C#
Assert.Equal(idx, jsonArray.Length);
```

kasiabulat replied:
I changed the tests in the meantime.

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+            var jsonArray = new JsonArray();
+
+            jsonArray.Insert(0, "value");
+            Assert.Equal("value", (JsonString)jsonArray[0]);
```
Length should be asserted.  A test should exist that takes an existing array, inserts somewhere not at the beginning or end, and that the contents afterwards look correct.  (i.e. verify that Insert isn't actually Set).

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+            jsonArray.Insert(0, true);
+            Assert.True(((JsonBoolean)jsonArray[0]).Value);
+
+            jsonArray.Insert(0, byte.MaxValue);
```
All of these insert at 0.  There should also be a set that inserts at the end. (Yeah, the first one is both 0 and the end, but this is "show that Length is special, not that 0 is special")

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+            jsonArray.Insert(0, short.MaxValue);
+            Assert.Equal(short.MaxValue, ((JsonNumber)jsonArray[0]).GetInt16());
+
+            jsonArray.Insert(0, int.MaxValue);
```
Negative tests are just as important as positive tests.  For each Add method there should be a test that inserting beyond Length or inserting negative numbers should throw an exception, and that exception should be validated (presumably ArgumentOutOfRangeException for a parameter named "index").

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+            jsonArray.Insert(0, decimal.One);
+            Assert.Equal(decimal.One, ((JsonNumber)jsonArray[0]).GetDecimal());
+        }
+
```
If Clone stays, add a test that confirms the deepness.  E.g.

```C#
JsonArray inner = new JsonArray { 1, 2, 3 };
JsonArray outer = new JsonArray { (JsonNode)inner) };
JsonArray outerClone = (JsonArray)outer.Clone();
((JsonArray)outerClone[0]).Add(4);

Assert.Equals(???, inner.Length);
```

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+            };
+
+            Assert.IsType<JsonArray>(preferences["colours"]);
+            var colours = preferences["colours"] as JsonArray;
```
```suggestion
            var colours = (JsonArray)preferences["colours"];
```

Use the hard cast when you aren't going to check for null.  (Default to hard cast)

bartonjs replied:
Actually, just combine the lines, IsType returns the value.

```C#
JsonArray colours = Assert.IsType<JsonArray>(preferences["colours"]);
```

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+        }
+
+        [Fact]
+        public static void TestCretingJsonArrayFromStringArray()
```
```suggestion
        public static void TestCreatingJsonArrayFromStringArray()
```

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+        }
+
+        [Fact]
+        public static void TestAddingJsonArray()
```
This feels more like a JsonObject test, since you're showing how JsonObject behaves.

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+        public static void TestCretingJsonArrayFromStringArray()
+        {
+            string[] expected = { "sushi", "pasta", "cucumber soup" };
+            var preferences = new JsonObject()
```
It looks like the purpose of this test is to test the JsonArray constructor.  JsonObject shouldn't be involved.

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+        {
+            var preferences = new JsonObject()
+            {
+                { "prime numbers", new JsonNumber[] { 19, 37 } }
```
This seems to be a JsonObject.Add(string, JsonNumbers) test, so it belongs in JsonObject tests.  (I'll stop pointing this out on this iteration)

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+            var innerJsonArray = vertices[0] as JsonArray;
+            Assert.IsType<JsonArray>(innerJsonArray[0]);
+            innerJsonArray = innerJsonArray[0] as JsonArray;
+            Assert.IsType<JsonArray>(innerJsonArray[0]);
```
Can we assert more about the contents?

bartonjs commented on src/System.Text.Json/tests/JsonArrayTests.cs:
```diff
@@ -0,0 +1,344 @@
+            Assert.Equal("feature 1569", (JsonString)((JsonArray)issues["features"])[0]);
+            Assert.Equal("feature 56134", (JsonString)((JsonArray)issues["features"])[1]);
+        }
+    }
```
* There are a lot of ctors, each one needs to be tested.
* All of the tests are homogeneous (always numbers or always strings), some heterogeneous tests should exist.
* Combine using the enumerable ctors and the add methods instead of always doing one or the other.

bartonjs commented on src/System.Text.Json/tests/JsonNodeTestData.cs:
```diff
@@ -0,0 +1,111 @@
+                "phone numbers",
+                new JsonObject()
+                {
+                    { "work", "123-456-7890" }, { "home", "123-456-7890" }
```
While this is an invalid phone number it's only because Nigerian phone numbers all have a "1" after the 4.  `425-555-01xx` is a better choice (`-555-01` is reserved)

kasiabulat replied:
Ok, I can change this...

bartonjs commented on src/System.Text.Json/tests/JsonNodeTransformationsTests.cs:
```diff
@@ -0,0 +1,187 @@
+        }
+
+        [Fact]
+        public static void TestAsJsonElement()
```
Might be worth splitting this into two different classes / purposes: JsonNodeToJsonElementTest and JsonNodeDeepCopyTests

kasiabulat commented:
I addressed review comments, made fixes and added documentation to `JsonNode`. I will add API (and specification) changes in new PR in case we want to come back to this version. I will change `JsonElement.ObjectEnumerator` not to use `JsonObjectEnumerator`, add more tests and improve existing ones in PR with API changes.

kasiabulat commented:
Do you think I can merge this PR?
